app/src/main/kotlin/com/mapbox/vision/teaser/MainActivity.kt

```diff
@@ -34,13 +62,108 @@ class MainActivity : BaseTeaserActivity() {
 
         VisionSafetyManager.create(VisionManager)
         VisionSafetyManager.visionSafetyListener = visionSafetyListener
+        return true
+    }
+
+    private fun initVisionManagerReplay(visionView: VisionView, sessionPath: String): Boolean {
+        if (sessionPath.isEmpty()) {
+            return false
+        }
+
+        this.sessionPath = sessionPath
+        VisionReplayManager.create(sessionPath)
+        VisionReplayManager.visionEventsListener = visionEventsListener
+        VisionReplayManager.start()
+        VisionReplayManager.setModelPerformance(modelPerformance)
+        visionView.setVisionManager(VisionReplayManager)
 
+        VisionSafetyManager.create(VisionReplayManager)
+        VisionSafetyManager.visionSafetyListener = visionSafetyListener
         return true
     }
 
+
+    private fun showReplayModeFragment() {
+        supportFragmentManager
+                .beginTransaction()
+                .replace(R.id.fragment_container, ReplayModeFragment.newInstance(BASE_SESSION_PATH), ReplayModeFragment.TAG)
+                .addToBackStack(ReplayModeFragment.TAG)
+                .commit()
+        hideDashboardView()
+
+    }
+
+    private fun hideDashboardView() {
+        dashboard_container.hide()
+        title_teaser.hide()
+    }
+
+    private fun showDashboardView() {
+        dashboard_container.show()
+        title_teaser.show()
+    }
+
     override fun destroyVisionManager() {
+        return when (mode) {
+            Realtime -> destroyVisionManagerRealtime()
+            Replay -> destroyVisionManagerReplay()
+            Recorder -> return
+        }
+    }
+
+    override fun getFrameStatistics(): FrameStatistics? {
+        return when (mode) {
+            Realtime -> VisionManager.getFrameStatistics()
+            Replay -> VisionReplayManager.getFrameStatistics()
+            Recorder -> null
+        }
+    }
+
+    private fun destroyVisionManagerRealtime() {
         VisionSafetyManager.destroy()
         VisionManager.stop()
         VisionManager.destroy()
     }
+
+    private fun destroyVisionManagerReplay() {
+        VisionSafetyManager.destroy()
+        VisionReplayManager.stop()
+        VisionReplayManager.destroy()
+    }
+
+    override fun onBackPressed() {
+        val fragment = supportFragmentManager.findFragmentById(R.id.fragment_container)
+        if (fragment is OnBackPressedListener) {
+            if (fragment.onBackPressed()) {
```

Comment:
```suggestion
        if (fragment is OnBackPressedListener && fragment.onBackPressed()) {
```

Reply:
They are not equal to each other, I guess. At the same time it's a question what's the correct behavior here, I think your suggestion is correct - if frament IS listening back presses but does NOT handle it we should pass it to super.

Reply:
It is not the same logic and 3 branches of execution for "onBackPressed"
1 - If you reset multiselection mode with back - Fragment handles onBack itself and reset multiselection mode, nothing happend from Activity side.
2 - Activity pops Fragment from BackStack if Fragment exists and no multiselection mode (Fragment doesn't handle this mode itself)
3 - Open Home screen if no Fragments, just regular "back"

Probably we can rewrite somehow this code better with your ideas, let's discuss better in Slack of with voice

Reply:
I've taken a fresh look and indeed @yuryybk's solution seems fine to me now.